Recent Kubernetes tarballs from GitHub don't build using make (wrong linker args?)

<!-- This form is for bug reports and feature requests ONLY! 

If you're looking for help check [Stack Overflow](https://stackoverflow.com/questions/tagged/kubernetes) and the [troubleshooting guide](https://kubernetes.io/docs/tasks/debug-application-cluster/troubleshooting/).
-->

/kind bug

**What happened**:
```
+++ [1124 14:16:45] Building the toolchain targets:
    k8s.io/kubernetes/hack/cmd/teststale
    k8s.io/kubernetes/vendor/github.com/jteeuwen/go-bindata/go-bindata
# k8s.io/kubernetes/vendor/github.com/jteeuwen/go-bindata/go-bindata
usage: link [options] main.o
(list of possible link args)
# k8s.io/kubernetes/hack/cmd/teststale
usage: link [options] main.o
(list of possible link args)
!!! [1124 14:16:46] Call tree:
!!! [1124 14:16:46]  1: /home/builder/aports/community/kubernetes/src/kubernetes-1.8.4/hack/lib/golang.sh:683 kube::golang::build_kube_toolchain(...)
!!! [1124 14:16:46]  2: hack/make-rules/build.sh:27 kube::golang::build_binaries(...)
!!! [1124 14:16:46] Call tree:
!!! [1124 14:16:46]  1: hack/make-rules/build.sh:27 kube::golang::build_binaries(...)
make[1]: *** [Makefile.generated_files:322: _output/local/go/bin/deepcopy-gen] Error 1
make: *** [Makefile:512: generated_files] Error 2
```

**What you expected to happen**:
Working build.

**How to reproduce it (as minimally and precisely as possible)**:
Run `make -j1` in Kubernetes 1.8.4 source unpacked from tarball.

**Anything else we need to know?**:
Kubernetes 1.8.3 tarball and git v1.8.4 tag build fine.
Link to tarball: https://github.com/kubernetes/kubernetes/archive/v1.8.4.tar.gz

**Environment**:
- Kubernetes version (use `kubectl version`): 1.8.4
- Cloud provider or hardware configuration:
- OS (e.g. from /etc/os-release): Alpine 3.6 and Ubuntu 17.10
- Kernel (e.g. `uname -a`): `Linux ubuntu-artful 4.13.0-16-generic #19-Ubuntu SMP Wed Oct 11 18:35:14 UTC 2017 x86_64 x86_64 x86_64 GNU/Linux`
- Install tools: Alpine: go 1.8.4; Ubuntu: go 1.8.3, 1.9.1
- Others:

--- a/build/lib/release.sh
+++ b/build/lib/release.sh
@@ -98,18 +98,23 @@ function kube::release::package_tarballs() {
 
 # Package the source code we built, for compliance/licensing/audit/yadda.
 function kube::release::package_src_tarball() {
+  local -r src_tarball="${RELEASE_TARS}/kubernetes-src.tar.gz"
   kube::log::status "Building tarball: src"
-  local source_files=(
-    $(cd "${KUBE_ROOT}" && find . -mindepth 1 -maxdepth 1 \
-      -not \( \
-        \( -path ./_\*        -o \
-           -path ./.git\*     -o \
-           -path ./.config\* -o \
-           -path ./.gsutil\*    \
-        \) -prune \
-      \))
-  )
-  "${TAR}" czf "${RELEASE_TARS}/kubernetes-src.tar.gz" -C "${KUBE_ROOT}" "${source_files[@]}"
+  if [[ "${KUBE_GIT_TREE_STATE-}" == "clean" ]]; then
+    git archive -o "${src_tarball}" HEAD
+  else
+    local source_files=(
+      $(cd "${KUBE_ROOT}" && find . -mindepth 1 -maxdepth 1 \
+        -not \( \
+          \( -path ./_\*        -o \
+             -path ./.git\*     -o \
+             -path ./.config\* -o \
+             -path ./.gsutil\*    \
+          \) -prune \
+        \))
+    )
+    "${TAR}" czf "${src_tarball}" -C "${KUBE_ROOT}" "${source_files[@]}"
+  fi
 }
 
 # Package up all of the cross compiled clients. Over time this should grow into

--- a/hack/lib/version.sh
+++ b/hack/lib/version.sh
@@ -22,6 +22,7 @@
 #          source code.
 #    KUBE_GIT_TREE_STATE - "clean" indicates no changes since the git commit id
 #        "dirty" indicates source code changes after the git commit id
+#        "archive" indicates the tree was produced by 'git archive'
 #    KUBE_GIT_VERSION - "vX.Y" used to indicate the last release version.
 #    KUBE_GIT_MAJOR - The major part of the version
 #    KUBE_GIT_MINOR - The minor component of the version
@@ -40,7 +41,7 @@ kube::version::get_version_vars() {
   # we likely don't have a git tree, but these magic values may be filled in.
   if [[ '$Format:%%$' == "%" ]]; then
     KUBE_GIT_COMMIT='$Format:%H$'
-    KUBE_GIT_TREE_STATE="git archive"
+    KUBE_GIT_TREE_STATE="archive"
     # When a 'git archive' is exported, the '$Format:%D$' below will look
     # something like 'HEAD -> release-1.8, tag: v1.8.3' where then 'tag: '
     # can be extracted from it.

